### PR TITLE
refactor(pd): 🍱 pull grpc endpoint into a standalone crate

### PR DESCRIPTION
**Pull request by cratelyn**
cherry-picked out of #3973. see #3913, and #3588.

we would like to use the rust view server in mock consensus tests. in order to run the `penumbra_view::ViewServer` however, we need to spin up the corresponding grpc endpoint for it to connect to.

this branch performs a bit of code motion, moving the `grpc_server` out of `pd` and into a standalone library that may be used within our mock consensus tests.

there will likely be other functional changes to the code in question before we can use it in those tests, but this PR is interested in moving that code into a place where our tests can rely upon it.

### CODE REVIEW

File: crates/bin/pd/src/main.rs [hunk 1]
```diff
@@ -134,102 +129,14 @@ async fn main() -> anyhow::Result<()> {
...
+                tonic_web::enable(TendermintProxyServiceServer::new(tm_proxy.clone())),
+            );
             if enable_expensive_rpc {
-                grpc_server = grpc_server.add_service(we(SimulationServiceServer::new(
-                    DexServer::new(storage.clone()),
-                )));
+                use penumbra_dex::component::rpc::Server as DexServer;
+                grpc_server = grpc_server.add_service(tonic_web::enable(
+                    SimulationServiceServer::new(DexServer::new(storage.clone())),
+                ));
```
Comment (cratelyn) [inline crates/bin/pd/src/main.rs:139]: for simplicity i left these both here. i don't have reason to believe yet that they will be needed by our tests, but they'd be easy to move over later if needed.

**Comment by cratelyn** [comment]
smoke tests are failing:

```
2024-03-08T21:19:33.340400Z ERROR penumbra_view::worker: view worker error e=status: Unknown, message: \"Service was not ready: transport error\", details: [], metadata: MetadataMap { headers: {} }
Error: status: Unknown, message: \"grpc-status header missing, mapped from HTTP status code 405\", details: [], metadata: MetadataMap { headers: {\"content-length\": \"0\", \"access-control-allow-origin\": \"*\", \"vary\": \"origin\", \"vary\": \"access-control-request-method\", \"vary\": \"access-control-request-headers\", \"access-control-expose-headers\": \"*\", \"allow\": \"GET,HEAD\", \"date\": \"Fri, 08 Mar 2024 21:19:32 GMT\"} }
```

i will track that down after the weekend. i believe i know how to fix that, i'll pick it up when i'm back on monday.

File: crates/bin/pd/src/main.rs (hunk 1 above)
Comment (cratelyn) [reply crates/bin/pd/src/main.rs:139]: https://github.com/penumbra-zone/penumbra/pull/3980/#issuecomment-1986436183

Comment (cratelyn) [reply crates/bin/pd/src/main.rs:139]: https://github.com/penumbra-zone/penumbra/pull/3980#issuecomment-1986436183

Comment (cratelyn) [reply crates/bin/pd/src/main.rs:139]: the order of `add_service` is load-bearing

**Comment by hdevalence** [comment]
Moving this code out of the `pd` crate seems good, but I'm wondering why we want to create a new crate, rather than putting it inside the `app` crate. Didn't we attempt to do that before, and ran into a Tonic bug? Can we avoid that bug while still moving the GRPC services into the `app` crate?

**Comment by cratelyn** [comment]
#### :clapper: re: deja vu

> Didn't we attempt to [move this] before, and ran into a Tonic bug? Can we avoid that bug while still moving the GRPC services into the `app` crate?

good memory, i am glad this sounds eerily familiar to you too! the difference is subtle.

#3679 was trying at the time to address [this](https://github.com/penumbra-zone/penumbra/pull/3679/files#diff-fbc4204ceb976c8cb30ed06168e2476700bae21bfd803e26281b2d026194d430L175-L182) `TODO` comment.

the difference between the two is that #3679 would have `penumbra-app` create a `Routes`, that we would [add](https://github.com/penumbra-zone/penumbra/pull/3679/files#diff-fbc4204ceb976c8cb30ed06168e2476700bae21bfd803e26281b2d026194d430R204) to the builder (_which stays in `pd`_).

here, i'm not trying to make that cut between `Router` and `Routes`, and am attempting to hoist the whole thing out of `pd`, without making changes to how we interact with `tonic`. my aim is for us to be able to move this, without running into that bug (#3697) again.

**Comment by cratelyn** [comment]
closing in favor of #3996! :clapper:

**Pull request closed without merging**